main/src/library/Applicative.flix

```diff
@@ -0,0 +1,83 @@
+/*
+ *  Copyright 2021 Felix Wiemuth
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+///
+/// A type class for functors that support application, i.e. allow to:
+/// - Make an applicative value out of a normal value (embed it into a default context), e.g. embed `5` into `Some(5)`.
+/// - Apply a function-type applicative to a matching argument-type applicative, resulting in an applicative of
+/// the function's result type. The meaning of this application is defined by the respective instance.
+/// Conceptually this can be understood as applying functions "contained" in the first applicative to arguments
+/// in the second applicative, where the possible quantity of functions/arguments depends on the type `m`.
+/// For example, an `Option[a -> b]` can be `None`, or contain a function of type `a -> b`, and only in the
+/// latter case a function is applied. A `List[a -> b]` contains a list of functions, which are all to be
+/// applied to all arguments in the arguments list.
+///
+/// A minimal implementation must define `point` and at least one of `ap` and `liftA2`. If it defines both,
+/// they must be equivalent to their default definitions:
+///     `ap(f: m[a -> b & e], x: m[a]) : m[b] & e = liftA2(identity, f, x)`
+///     `liftA2(f: a -> b -> c & e, x: m[a], y: m[b]): m[c] & e = ap(Functor.map(f, x), y)`
+///
```

Comment:
This of course means that you can forget to define either. What do you think about that?

Reply:
Yes that is of course a problem: if none is implemented, calling either will lead to non-termination.

For now I think we cannot do much more than document it if we want the flexibility of allowing to implement either one of the functions.

In the future we could introduce an advanced mechanism for type classes allowing to specify which functions have to be implemented. This could be a boolean expression composed of variables corresponding to the to-be-implemented functions. In this case the expression would be `ap v liftA2`.

Reply:
> In the future we could introduce an advanced mechanism for type classes allowing to specify which functions have to be implemented. This could be a boolean expression composed of variables corresponding to the to-be-implemented functions. In this case the expression would be `ap v liftA2`.

I like this idea.


But for now, how about we leave `ap` unimplemented? 

You can still write an instance that defines `ap` in terms of `lift` and overrides lift.

Reply:
> But for now, how about we leave ap unimplemented?
> You can still write an instance that defines ap in terms of lift and overrides lift.

Right, that sounds like the best option for now. Usually it is `ap` one wants to implement, and if one wants to implement `liftA2` instead, one can implement `ap` based on it like the default implementation does now. I could add that to the documentation.

Reply:
Agree. We do the same for e.g. `Order` where you must implement `compare`.